torch/csrc/jit/autodiff.cpp

```diff
@@ -106,6 +107,19 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
     } else if (node->matches("aten::relu(Tensor self) -> Tensor")) {
       return {grads.at(0) * (outputs.at(0) > at::Scalar(0)).type_as(outputs.at(0))};
 
+    } else if (node->matches("aten::clamp(Tensor self, Scalar min, Scalar max) -> Tensor")) {
+      // we do two type_as as it's free (hopefully) and the "*" only works with float
+      // the "! (val > min)" is chosen such that the gradient is 0 on the
+      // boundary and the factor is 1 when the boundary is NaN
+      // the ! is expressed as "1-" for lack of a "not" function and
+      // the the fuser insisting on float
+      // it would be prettier to return NaN as gradient for NaN,
+      // but that is hard to reliably code here, so we have 0 as gradient
+      // when the input is NaN (unless grads is NaN or infinite)
+      return {grads.at(0)
+	      * (1-(inputs.at(0).isnan()).type_as(inputs.at(0)))
```

Comment:
This term is missing from the derivative formula we usually use for clamp. I'm not sure why do we want it. I don't think it's worth sacrifising performance just to change the NaN handling.

Reply:
Personally, I think that proper NaN handling is a good feature and was about to fix it, but I'll take it out if the policy is not to treat NaN. (I do seem to recall from the forum that it caused quite bit of confusion that ReLU forward used to swallow NaN.)